primer/src/Primer/Action.hs

```diff
@@ -910,35 +968,72 @@ toProgActionNoInput defs def defName mNodeSel = \case
   Available.MakeTApp ->
     toProgAction [ConstructTApp, Move Child1]
   Available.RaiseType -> do
-    id <- mid
-    pure [MoveToDef defName, CopyPasteSig (defName, id) [SetCursor id, Move Parent, Delete]]
+    id <- nodeID
+    sel <- termSel
+    pure [MoveToDef sel.def, CopyPasteSig (sel.def, id) [SetCursor id, Move Parent, Delete]]
   Available.DeleteType ->
     toProgAction [Delete]
-  Available.DuplicateDef ->
+  Available.DuplicateDef -> do
+    sel <- termSel
+    def <- termDef
     let sigID = getID $ astDefType def
         bodyID = getID $ astDefExpr def
-        copyName = uniquifyDefName (qualifiedModule defName) (unName (baseName defName) <> "Copy") defs
+        copyName = uniquifyDefName (qualifiedModule sel.def) (unName (baseName sel.def) <> "Copy") defs
      in pure
-          [ CreateDef (qualifiedModule defName) (Just copyName)
-          , CopyPasteSig (defName, sigID) []
-          , CopyPasteBody (defName, bodyID) []
+          [ CreateDef (qualifiedModule sel.def) (Just copyName)
+          , CopyPasteSig (sel.def, sigID) []
+          , CopyPasteBody (sel.def, bodyID) []
           ]
-  Available.DeleteDef ->
-    pure [DeleteDef defName]
+  Available.DeleteDef -> do
+    sel <- termSel
+    pure [DeleteDef sel.def]
+  Available.AddConField -> do
+    (defName, sel) <- conSel
+    d <- typeDef
+    vc <-
+      maybe (Left $ ValConNotFound defName sel.con) pure
+        . find ((== sel.con) . valConName)
+        $ astTypeDefConstructors d
+    let index = length $ valConArgs vc -- for now, we always add on to the end
+    pure [AddConField defName sel.con index $ TEmptyHole ()]
   where
-    toProgAction actions = toProg' actions defName <$> maybeToEither NoNodeSelection mNodeSel
-    mid = maybeToEither NoNodeSelection $ snd <$> mNodeSel
+    termSel = case sel0 of
+      SelectionDef s -> pure s
+      SelectionTypeDef _ -> Left NeedTermDefSelection
+    nodeID = do
+      sel <- termSel
+      maybeToEither NoNodeSelection $ (.meta) <$> sel.node
+    typeSel = case sel0 of
+      SelectionDef _ -> Left NeedTypeDefSelection
+      SelectionTypeDef s -> pure s
+    typeNodeSel = do
+      sel <- typeSel
+      maybe (Left NeedTypeDefNodeSelection) (pure . (sel.def,)) sel.node
+    conSel =
+      typeNodeSel >>= \case
+        (s0, TypeDefConsNodeSelection s) -> pure (s0, s)
+        _ -> Left NeedTypeDefConsSelection
+    conFieldSel = do
+      (ty, s) <- conSel
+      maybe (Left NeedTypeDefConsFieldSelection) (pure . (ty,s.con,)) s.field
+    toProgAction actions = do
+      case sel0 of
+        SelectionDef sel -> toProg' actions sel.def <$> maybeToEither NoNodeSelection sel.node
+        SelectionTypeDef _ -> do
+          (t, c, f) <- conFieldSel
+          pure [ConFieldAction t c f.index $ SetCursor f.meta : actions]
+    termDef = first (const NeedTermDef) def0
+    typeDef = either Right (Left . const NeedTypeDef) def0
```

Comment:
A lot of these helpers also occur in `toProgActionInput`. I started trying to clean this up in #1041, but got in to the weeds a little. I'd prefer that we avoid this repetition for this PR, but revisit soon.

Reply:
> I'd prefer that we avoid this repetition for this PR, but revisit soon.

As in: let's just commit this as-is, and improve it in #1041?

Reply:
Ah, yes, I seemed to be missing a few words.